### PR TITLE
common: wasmToSchema: Base custom section resolution on module version

### PR DESCRIPTION
**Pull request by bisgardo**
## Purpose

Ensure that `getEmbeddedSchema` (on `ConcordiumGRPCClient`) is able to resolve legacy schemas (i.e. schemas without embedded versions).

Also make parsing of contract schema more precise based by using the module version information that is now exposed by the gRPC client. This in turn also documents the relationship between module and schema versions.

## Changes

The module version is now used to determine what custom sections to look for. Added comments explain the expected results.

The extraction code is migrated from the dapp-library [`@concordium/react-components`](https://github.com/Concordium/concordium-dapp-libraries/blob/a045f3b8bc11c7389190c789072f313686b6c639/packages/react-components/src/useModuleSchemaRpc.ts). Specifically, the contents of function `getEmbeddedModuleSchema` was replaced with the contents of `findSchema` and its helper function `findCustomSections` was added and adjusted to use the SDK types.

Finally, `getEmbeddedModuleSchema` was made non-async and `wasmToSchema` removed because it was nothing but a more limited version of `getEmbeddedModuleSchema`.

Addresses https://github.com/Concordium/concordium-dapp-libraries/issues/66.

### CODE REVIEW

**Comment by bisgardo** [comment]
@orhoj @shjortConcordium @rasmus-kirk @abizjak Do we want this? If so I'll clean it up and add some tests.

I think it would be useful and simplify https://github.com/Concordium/concordium-dapp-libraries/pull/46/files.

**Comment by soerenbf** [comment]
@bisgardo Stumbled upon this PR with no reviewers and no replies. Is this still relevant? I guess if it's a general purpose thing that you've needed to implement in dapp-libraries, it might as well exist here instead?

I'm sorry this has been dangling for so long. If you deem it relevant still, I'll be happy to take a look if you get it up to speed with the main branch.

Otherwise please close it 😄

**Comment by bisgardo** [comment]
> @bisgardo Stumbled upon this PR with no reviewers and no replies. Is this still relevant? I guess if it's a general purpose thing that you've needed to implement in dapp-libraries, it might as well exist here instead?
> 
> I'm sorry this has been dangling for so long. If you deem it relevant still, I'll be happy to take a look if you get it up to speed with the main branch.
> 
> Otherwise please close it 😄

@soerenbf Thanks for poking at this. I still think it's relevant :) The reason I didn't add reviewers was that I didn't consider it ready for review and I didn't want to finish it up before it was deemed worthy of doing.

I merged main into the PR branch. A couple of things have changed in the meantime and the feature seems more widely used and there seems to be some duplication that I'll try and address as part of it.

File: packages/sdk/src/types/VersionedModuleSource.ts
Comment (soerenbf) [inline packages/sdk/src/types/VersionedModuleSource.ts:121]: It would be nice to spell this out explicitly, i.e. `!== undefined ...`

File: packages/sdk/src/pub/types.ts
Comment (soerenbf) [inline packages/sdk/src/pub/types.ts:28]: This is a breaking change

File: packages/sdk/src/grpc/GRPCClient.ts
Comment (soerenbf) [inline packages/sdk/src/grpc/GRPCClient.ts:238]: This is a breaking change

File: packages/sdk/CHANGELOG.md
Comment (soerenbf) [inline packages/sdk/CHANGELOG.md:34]: These are breaking changes and should be declared as such

File: packages/sdk/src/types/VersionedModuleSource.ts
Comment (soerenbf) [inline packages/sdk/src/types/VersionedModuleSource.ts:116]: Also breaking

**Review by soerenbf — COMMENTED** [review]
As these are breaking changes, I don't think these should be just merged to `main`. As I see it, we have 2 options: either change these to not be breaking (if possible), or follow the process outlined below.

The way we usually handle breaking changes is:

- Make a new release branch, i.e. `release/web-sdk/8`
- Target this branch for PR's implementing breaking changes
- Merge this when we're ready.

If you need to use these changes, I would suggest publishing them to npm under either the `alpha` or `beta` tag (we usually use `alpha` for internal preliminary release versions. If you intend to publish the `dapp-libraries` based on this, I would use `beta`).

**Comment by bisgardo** [comment]
> As these are breaking changes, I don't think these should be just merged to `main`. As I see it, we have 2 options: either change these to not be breaking (if possible), or follow the process outlined below.
> 
> The way we usually handle breaking changes is:
> 
>     * Make a new release branch, i.e. `release/web-sdk/8`
> 
>     * Target this branch for PR's implementing breaking changes
> 
>     * Merge this when we're ready.
> 
> 
> If you need to use these changes, I would suggest publishing them to npm under either the `alpha` or `beta` tag (we usually use `alpha` for internal preliminary release versions. If you intend to publish the `dapp-libraries` based on this, I would use `beta`).

I'm in favor of keeping it a breaking change because having `ConcordiumGRPCClient.getEmbeddedSchema` only support new kinds of schemas (and actually throw an error if it's an old one) doesn't make a whole lot of sense. I would actually go a step further and remove it entirely because it's trivially expressed in terms of `getModuleSource` and `getEmbeddedModuleSchema`.

I'll let you decide what should happen. Just let me know and I'll make the relevant changes.

Btw. I'm in no rush to use this in dapp-libs. But I'm happy to prepare a PR to be merged when the time is right.

File: packages/sdk/src/types/VersionedModuleSource.ts
Comment (bisgardo) [reply packages/sdk/src/types/VersionedModuleSource.ts]: done

File: packages/sdk/src/grpc/GRPCClient.ts
Comment (bisgardo) [reply packages/sdk/src/grpc/GRPCClient.ts:238]: Already mentioned in changelog.

File: packages/sdk/src/types/VersionedModuleSource.ts
Comment (bisgardo) [reply packages/sdk/src/types/VersionedModuleSource.ts:116]: Already mentioned in changelog.

File: packages/sdk/CHANGELOG.md
Comment (bisgardo) [reply packages/sdk/CHANGELOG.md]: done

File: packages/sdk/src/pub/types.ts
Comment (bisgardo) [reply packages/sdk/src/pub/types.ts:28]: Already mentioned in changelog.

File: packages/sdk/src/types/VersionedModuleSource.ts
Comment (bisgardo) [inline packages/sdk/src/types/VersionedModuleSource.ts:116]: While we're at it we could also change it to return `undefined` instead of `null`? It seems to be an outlier in returning `null`.

Comment (soerenbf) [reply packages/sdk/src/types/VersionedModuleSource.ts:116]: Sure, better to be consistent.

**Comment by soerenbf** [comment]
> I'm in favor of keeping it a breaking change because having `ConcordiumGRPCClient.getEmbeddedSchema` only support new kinds of schemas (and actually throw an error if it's an old one) doesn't make a whole lot of sense. I would actually go a step further and remove it entirely because it's trivially expressed in terms of `getModuleSource` and `getEmbeddedModuleSchema`.
> 
> I'll let you decide what should happen. Just let me know and I'll make the relevant changes.

I think I would prefer to just keep the method on the GRPC client.

Please create a release branch for the next major version of the SDK and set it as the target branch for this. Then we'll release the changes with the next protocol update!

**Review by soerenbf — APPROVED** [review]
Assuming this is _not_ merged to main but to a release branch for the next major version 😉

File: packages/sdk/src/types/VersionedModuleSource.ts
Comment (bisgardo) [reply packages/sdk/src/types/VersionedModuleSource.ts:116]: done

**Pull request merged**
